File: plugins/tagManager/tests/test_blacklist.py
"""
Unit tests for blacklist module.
Run with: python -m pytest plugins/tagManager/tests/test_blacklist.py -v
"""

import pytest
import sys
import os

# Add parent directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from blacklist import Blacklist


class TestBlacklistParsing:
    """Test blacklist pattern parsing."""

    def test_empty_blacklist(self):
        bl = Blacklist(None)
        assert bl.count == 0

    def test_empty_string(self):
        bl = Blacklist('')
        assert bl.count == 0

    def test_literal_patterns(self):
        bl = Blacklist('4K Available\nFull HD Available')
        assert bl.count == 2
        assert len(bl.literals) == 2
        assert len(bl.regexes) == 0

    def test_regex_patterns(self):
        bl = Blacklist('/^\\d+p$\n/Available$')
        assert bl.count == 2
        assert len(bl.literals) == 0
        assert len(bl.regexes) == 2

    def test_mixed_patterns(self):
        bl = Blacklist('4K Available\n/Available$')
        assert bl.count == 2
        assert len(bl.literals) == 1
        assert len(bl.regexes) == 1

    def test_invalid_regex_skipped(self):
        bl = Blacklist('/[invalid/')
        assert bl.count == 0  # Invalid regex should be skipped

    def test_blank_lines_ignored(self):
        bl = Blacklist('Pattern1\n\n\nPattern2\n  \n')
        assert bl.count == 2


class TestBlacklistMatching:
    """Test tag name matching."""

    def test_literal_exact_match(self):
        bl = Blacklist('4K Available')
        assert bl.is_blacklisted('4K Available') is True

    def test_literal_case_insensitive(self):
        bl = Blacklist('4K Available')
        assert bl.is_blacklisted('4k available') is True
        assert bl.is_blacklisted('4K AVAILABLE') is True

    def test_literal_no_partial_match(self):
        bl = Blacklist('4K')
        assert bl.is_blacklisted('4K Available') is False

    def test_regex_match(self):
        bl = Blacklist('/Available$')
        assert bl.is_blacklisted('4K Available') is True
        assert bl.is_blacklisted('Full HD Available') is True
        assert bl.is_blacklisted('Available Now') is False  # Not at end

    def test_regex_resolution_pattern(self):
        bl = Blacklist('/^\\d+p$')
        assert bl.is_blacklisted('1080p') is True
        assert bl.is_blacklisted('720p') is True
        assert bl.is_blacklisted('1080p Video') is False

    def test_non_blacklisted_tag(self):
        bl = Blacklist('4K Available')
        assert bl.is_blacklisted('Action') is False

    def test_empty_tag_name(self):
        bl = Blacklist('Pattern')
        assert bl.is_blacklisted('') is False
        assert bl.is_blacklisted(None) is False


class TestBlacklistFilter:
    """Test tag list filtering."""

    def test_filter_dict_tags(self):
        bl = Blacklist('4K Available')
        tags = [
            {'name': 'Action'},
            {'name': '4K Available'},
            {'name': 'Comedy'}
        ]
        filtered, hidden = bl.filter_tags(tags)
        assert len(filtered) == 2
        assert hidden == 1
        assert all(t['name'] != '4K Available' for t in filtered)

    def test_filter_empty_blacklist(self):
        bl = Blacklist('')
        tags = [{'name': 'Action'}, {'name': 'Comedy'}]
        filtered, hidden = bl.filter_tags(tags)
        assert len(filtered) == 2
        assert hidden == 0

    def test_filter_all_blacklisted(self):
        bl = Blacklist('Action\nComedy')
        tags = [{'name': 'Action'}, {'name': 'Comedy'}]
        filtered, hidden = bl.filter_tags(tags)
        assert len(filtered) == 0
        assert hidden == 2

File: plugins/tagManager/tests/test_blacklist.js
/**
 * Unit tests for blacklist functions.
 * Run with: node plugins/tagManager/tests/test_blacklist.js
 */

// Copy of parseBlacklist for testing
function parseBlacklist(blacklistStr) {
  if (!blacklistStr) return [];

  return blacklistStr.split('\n')
    .map(line => line.trim())
    .filter(line => line.length > 0)
    .map(pattern => {
      if (pattern.startsWith('/')) {
        const regexStr = pattern.slice(1);
        try {
          return { type: 'regex', pattern: regexStr, regex: new RegExp(regexStr, 'i') };
        } catch (e) {
          return null;
        }
      } else {
        return { type: 'literal', pattern: pattern.toLowerCase() };
      }
    })
    .filter(p => p !== null);
}

// Copy of isBlacklisted for testing
let tagBlacklist = [];

function isBlacklisted(tagName) {
  if (!tagName || tagBlacklist.length === 0) return false;

  const lowerName = tagName.toLowerCase();

  for (const entry of tagBlacklist) {
    if (entry.type === 'literal') {
      if (lowerName === entry.pattern) return true;
    } else if (entry.type === 'regex') {
      if (entry.regex.test(tagName)) return true;
    }
  }

  return false;
}

// Test runner
let passed = 0;
let failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`✓ ${name}`);
    passed++;
  } catch (e) {
    console.log(`✗ ${name}`);
    console.log(`  Error: ${e.message}`);
    failed++;
  }
}

function assertEqual(actual, expected, msg = '') {
  if (JSON.stringify(actual) !== JSON.stringify(expected)) {
    throw new Error(`${msg}\n  Expected: ${JSON.stringify(expected)}\n  Actual: ${JSON.stringify(actual)}`);
  }
}

// Tests
console.log('\n=== parseBlacklist tests ===\n');

test('returns empty array for null/undefined', () => {
  assertEqual(parseBlacklist(null).length, 0);
  assertEqual(parseBlacklist(undefined).length, 0);
  assertEqual(parseBlacklist('').length, 0);
});

test('parses literal patterns', () => {
  const result = parseBlacklist('4K Available\nFull HD');
  assertEqual(result.length, 2);
  assertEqual(result[0].type, 'literal');
  assertEqual(result[0].pattern, '4k available');
});

test('parses regex patterns', () => {
  const result = parseBlacklist('/Available$');
  assertEqual(result.length, 1);
  assertEqual(result[0].type, 'regex');
  assertEqual(result[0].pattern, 'Available$');
});

test('skips invalid regex', () => {
  const result = parseBlacklist('/[invalid/');
  assertEqual(result.length, 0);
});

test('ignores blank lines', () => {
  const result = parseBlacklist('Pattern1\n\n\nPattern2');
  assertEqual(result.length, 2);
});

console.log('\n=== isBlacklisted tests ===\n');

test('literal exact match (case-insensitive)', () => {
  tagBlacklist = parseBlacklist('4K Available');
  assertEqual(isBlacklisted('4K Available'), true);
  assertEqual(isBlacklisted('4k available'), true);
  assertEqual(isBlacklisted('Action'), false);
});

test('regex pattern matching', () => {
  tagBlacklist = parseBlacklist('/Available$');
  assertEqual(isBlacklisted('4K Available'), true);
  assertEqual(isBlacklisted('Full HD Available'), true);
  assertEqual(isBlacklisted('Available Now'), false);
});

test('resolution pattern', () => {
  tagBlacklist = parseBlacklist('/^\\d+p$');
  assertEqual(isBlacklisted('1080p'), true);
  assertEqual(isBlacklisted('720p'), true);
  assertEqual(isBlacklisted('1080p Video'), false);
});

test('empty name returns false', () => {
  tagBlacklist = parseBlacklist('Pattern');
  assertEqual(isBlacklisted(''), false);
  assertEqual(isBlacklisted(null), false);
});

test('empty blacklist returns false', () => {
  tagBlacklist = [];
  assertEqual(isBlacklisted('Anything'), false);
});

// Summary
console.log('\n=== Summary ===\n');
console.log(`Passed: ${passed}`);
console.log(`Failed: ${failed}`);

if (failed > 0) {
  process.exit(1);
}
